fala/apps/adviser/views.py

```diff
@@ -37,3 +38,114 @@ def get(self, request, *args, **kwargs):
 
 class AccessibilityView(TemplateView):
     template_name = "adviser/accessibility-statement.html"
+
+
+class SearchView(ListView):
+    class ErrorState(object):
+        def __init__(self, form):
+            self._form = form
+
+        @property
+        def template_name(self):
+            return "search.html"
+
+        def get_queryset(self):
+            return []
+
+        def get_context_data(self):
+            return {"form": self._form, "data": {}}
+
+    class EnglandOrWalesState(object):
+        def __init__(self, form):
+            self._form = form
+            self._data = form.search()
+
+        @property
+        def template_name(self):
+            return "results.html"
+
+        def get_queryset(self):
+            return self._data.get("results", None)
+
+        def get_context_data(self):
+            pages = LaaLaaPaginator(self._data["count"], 10, 3, self._form.current_page)
+            params = {
+                "postcode": self._form.cleaned_data["postcode"],
+                "name": self._form.cleaned_data["name"],
+            }
+            # create list of tuples which can be passed to urlencode for pagination links
+            categories = [("categories", c) for c in self._form.cleaned_data["categories"]]
+            return {
+                "form": self._form,
+                "data": self._data,
+                "pages": pages,
+                "params": params,
+                "categories": categories,
+            }
+
+    class OtherJurisdictionState(object):
+        REGION_TO_LINK = {
+            Region.NI: {
+                "link": "https://www.nidirect.gov.uk/articles/legal-aid-schemes",
+                "region": "Northern Ireland",
+            },
+            Region.IOM: {
+                "link": "https://www.gov.im/categories/benefits-and-financial-support/legal-aid/",
+                "region": "Isle of Man",
+            },
+            Region.JERSEY: {
+                "link": "https://www.legalaid.je/",
+                "region": "Jersey",
+            },
+            Region.GUERNSEY: {
+                "link": "https://www.gov.gg/legalaid",
+                "region": "Guernsey",
+            },
+            Region.SCOTLAND: {
+                "link": "https://www.mygov.scot/legal-aid/",
+                "region": "Scotland",
+            },
+        }
+
+        def __init__(self, region, postcode):
+            self._region = region
+            self._postcode = postcode
+
+        def get_queryset(self):
+            return []
+
+        @property
+        def template_name(self):
+            return "other_region.html"
+
+        def get_context_data(self):
+            region_data = self.REGION_TO_LINK[self._region]
```

Comment:
OK between the consolidation of related -- and interdependent -- logic into states and me reading through the form class and the view code to understand that we can guarantee that `self._region` exists in `self.REGION_TO_LINK`, I'm now no longer worried about this causing an exception.
I don't super love it as it's far from self-evident why it's not as risky as it looks, but it's good to go.